File: android/app/src/main/kotlin/in/nann/smashhockey/ui/Slider3D.kt
package `in`.nann.smashhockey.ui

import `in`.nann.smashhockey.ui.DesignTokens.Colour
import `in`.nann.smashhockey.ui.DesignTokens.Size
import kotlin.math.max
import kotlin.math.min

/**
 * A puck on a rail, for the coach's board values 0–1 — the twin of iOS's `Slider3D`. Grab it
 * anywhere on the row and drag; the puck chases the finger on the `bouncy` spring, leans into the
 * direction it travels and squashes while held. The fill behind it shows the value. TalkBack
 * adjusts it in steps.
 */
class Slider3D(
    private val kit: Kit,
    title: String,
    id: String,
    value: Double,
    val length: Float = 1.4f,
    private val step: Double = 0.05,
    entrance: Entrance = Entrance.Slide(fromLeft = true),
) : Interactive, Presentable {
    override val node = kit.node(null)
    private val body = kit.node(node)
    private val fill: UiNode
    private val puck = kit.node(body)
    private val readout: UiNode
    override val semantics = Semantics(id, title, percent(value), Semantics.Trait.ADJUSTABLE)
    val rest = Xform()
    val presence = Presence(entrance, kit.motion)
    var value = value
        private set
    var onChange: ((Double) -> Unit)? = null
    private val motion = kit.motion
    private val knob = Spring(motion.bouncy, value)
    private val grab = Spring(motion.bouncy)
    private var held = false
    private val rowHeight = 0.34f

    init {
        val railH = Size.RAIL_HEIGHT
        kit.slab(length + railH, railH, railH, Colour.RAIL, body, corner = railH / 2)
        fill = kit.slab(1f, railH * 1.25f, railH * 1.25f, Colour.SUN, body, corner = 0f)
        val r = Size.KNOB_RADIUS
        val upright = Quat().axisAngle((Math.PI / 2).toFloat(), 1f, 0f, 0f)
        kit.cylinder(Size.KNOB_DEPTH, r, Colour.INK, puck).setRotation(upright)
        val cap = kit.cylinder(Size.KNOB_DEPTH * 0.3f, r * 0.55f, Colour.CREAM, puck)
        cap.setRotation(upright)
        cap.setPosition(0f, 0f, Size.KNOB_DEPTH * 0.5f)
        puck.setPosition(0f, 0f, railH)

        val titleNode = kit.node(body)
        val readoutNode = kit.node(body)
        val titleText = kit.text(title, Size.TEXT_BODY, Colour.CREAM, titleNode)
        readout = kit.text(percent(value), Size.TEXT_BODY, Colour.SUN, readoutNode)
        titleNode.setPosition(-length / 2 + kit.width(titleText) / 2, 0.12f, 0f)
        readoutNode.setPosition(length / 2 - 0.12f, 0.12f, 0f)
        node.enabled = false
    }

    override val boundsNode get() = node
    override val bounds = Bounds(-length / 2 - 0.1f, -rowHeight / 2 + 0.02f, -0.1f, length / 2 + 0.1f, rowHeight / 2, 0.15f)
    override val isPresent get() = presence.isSettledIn

    override fun show(after: Double) = presence.show(after)
    override fun hide(after: Double) { held = false; presence.hide(after) }

    fun set(v: Double, notify: Boolean = true) {
        val q = min(1.0, max(0.0, (v / 0.01).roundHalfAway() * 0.01))
        if (q == value) return
        value = q
        knob.target = q
        semantics.value = percent(q)
        kit.retext(readout, percent(q), Size.TEXT_BODY)
        if (notify) onChange?.invoke(q)
    }

    private fun follow(ray: TouchRay) {
        val x = ray.planeX() ?: return     // the stage hands rays in our own space
        set(((x + length / 2) / length).toDouble())
    }

    override fun touchDown(ray: TouchRay) {
        held = true
        grab.target = 1.0
        grab.kick(motion.kick(KickName.GRAB))
        follow(ray)
    }

    override fun touchMoved(ray: TouchRay) { if (held) follow(ray) }

    override fun touchUp(ray: TouchRay, inside: Boolean) {
        held = false
        grab.target = 0.0
    }

    override fun activate() = grab.kick(motion.kick(KickName.GRAB))

    override fun adjust(steps: Int) {
        set(((value + steps * step) / step).roundHalfAway() * step)
        grab.kick(motion.kick(KickName.GRAB))
    }

    override fun update(dt: Double, ctx: UiContext) {
        presence.advance(dt, ctx)
        presence.apply(node, rest, ctx.reduceMotion)
        if (!node.enabled) return
        if (ctx.reduceMotion) {
            knob.snap(value)
            grab.snap(if (held) 1.0 else 0.0)
        } else {
            knob.advance(dt)
            grab.advance(dt)
        }
        val k = min(1.08, max(-0.08, knob.value)).toFloat()
        val x = -length / 2 + k * length
        val g = grab.value.toFloat()
        // Lean into the travel, squash while held.
        val lean = max(-0.5, min(0.5, knob.velocity * 0.12)).toFloat()
        puck.transform.tx = x
        puck.transform.rot.axisAngle(-lean, 0f, 0f, 1f)
        puck.setScale(1 + 0.25f * g, 1 + 0.25f * g, 1 - 0.3f * g)
        val w = max(0.001f, x + length / 2)
        fill.transform.sx = w
        fill.transform.tx = -length / 2 + w / 2
        fill.changed()
    }

    private companion object {
        fun percent(v: Double) = "${(v * 100).roundHalfAway().toInt()}%"
        /** Swift's `rounded()`: half away from zero (Kotlin's `round` is half-even). */
        fun Double.roundHalfAway(): Double = if (this >= 0) kotlin.math.floor(this + 0.5) else -kotlin.math.floor(-this + 0.5)
    }
}

File: ios/Sources/UI/Slider3D.swift
import Foundation
import RealityKit

/// A puck on a rail, for the coach's board values 0–1. Grab it anywhere on the row and drag; the
/// puck chases the finger on the `bouncy` spring, leans into the direction it travels and
/// squashes while held. The fill behind it shows the value. VoiceOver adjusts it in steps.
/// With `stops`, the value clicks between that many evenly spaced positions (the board's period
/// length and ball spin, §12), and `format` writes the readout.
@MainActor
final class Slider3D: Interactive, Presentable {
    let entity = Entity()
    private let body = Entity()
    private let fill: ModelEntity
    private let puck = Entity()
    private let title: ModelEntity
    private let readout: ModelEntity
    private(set) var semantics: Semantics
    var rest = Transform()
    var presence: Presence
    private(set) var value: Double
    var onChange: ((Double) -> Void)?
    let length: Float
    private let step: Double
    private let stops: Int?
    private let format: (Double) -> String
    private var knob: Spring
    private var grab: Spring
    private var held = false
    private let motion: MotionTokens
    private let rowHeight: Float = 0.34

    init(_ title: String, id: String, value: Double, length: Float = 1.4, step: Double = 0.05, stops: Int? = nil,
         format: ((Double) -> String)? = nil, entrance: Entrance = .slide(fromLeft: true), motion: MotionTokens) {
        self.value = value
        self.length = length
        self.stops = stops
        self.step = stops.map { 1 / Double($0 - 1) } ?? step
        self.format = format ?? Self.percent
        self.motion = motion
        semantics = Semantics(id: id, label: title, value: self.format(value), trait: .adjustable)
        presence = Presence(entrance, motion: motion)
        knob = Spring(motion.bouncy, initial: value)
        grab = Spring(motion.bouncy)

        let railH = DesignTokens.Size.railHeight
        let rail = Blocks.slab([length + railH, railH, railH], DesignTokens.Colour.rail, corner: railH / 2)
        fill = Blocks.slab([1, railH * 1.25, railH * 1.25], DesignTokens.Colour.sun, corner: 0.0)
        let r = DesignTokens.Size.knobRadius
        let disc = Blocks.model(.generateCylinder(height: DesignTokens.Size.knobDepth, radius: r), DesignTokens.Colour.ink)
        disc.orientation = simd_quatf(angle: .pi / 2, axis: [1, 0, 0])
        let cap = Blocks.model(.generateCylinder(height: DesignTokens.Size.knobDepth * 0.3, radius: r * 0.55), DesignTokens.Colour.cream)
        cap.orientation = disc.orientation
        cap.position.z = DesignTokens.Size.knobDepth * 0.5
        puck.addChild(disc)
        puck.addChild(cap)
        puck.position.z = railH

        self.title = Blocks.text(title, height: DesignTokens.Size.textBody, DesignTokens.Colour.cream)
        readout = Blocks.text(self.format(value), height: DesignTokens.Size.textBody, DesignTokens.Colour.sun)
        let titleNode = Entity(), readoutNode = Entity()
        titleNode.addChild(self.title)
        readoutNode.addChild(readout)
        titleNode.position = [-length / 2 + Blocks.width(of: self.title) / 2, 0.12, 0]
        readoutNode.position = [length / 2 - 0.12, 0.12, 0]

        entity.addChild(body)
        body.addChild(rail)
        body.addChild(fill)
        body.addChild(puck)
        body.addChild(titleNode)
        body.addChild(readoutNode)
        entity.isEnabled = false
    }

    private static func percent(_ v: Double) -> String { "\(Int((v * 100).rounded()))%" }

    var boundsEntity: Entity { entity }
    var bounds: BoundingBox {
        BoundingBox(min: [-length / 2 - 0.1, -rowHeight / 2 + 0.02, -0.1], max: [length / 2 + 0.1, rowHeight / 2, 0.15])
    }
    var isPresent: Bool { presence.isSettledIn }

    func show(after delay: Double) { presence.show(after: delay) }
    func hide(after delay: Double) { held = false; presence.hide(after: delay) }

    func set(_ v: Double, notify: Bool = true) {
        let c = min(1, max(0, v))
        let q = stops.map { n in (c * Double(n - 1)).rounded() / Double(n - 1) } ?? (c / 0.01).rounded() * 0.01
        guard q != value else { return }
        if notify, (q / step).rounded(.down) != (value / step).rounded(.down) { KitSound.step() }       // a step crossed
        value = q
        knob.target = q
        semantics.value = format(q)
        Blocks.retext(readout, format(q), height: DesignTokens.Size.textBody)
        if notify { onChange?(q) }
    }

    private func follow(_ ray: TouchRay) {
        guard let p = ray.onPlane else { return }   // the stage hands rays in our own space
        set(Double((p.x + length / 2) / length))
    }

    func touchDown(_ ray: TouchRay) {
        held = true
        grab.target = 1
        grab.kick(motion.kick(.grab))
        follow(ray)
    }

    func touchMoved(_ ray: TouchRay) { if held { follow(ray) } }

    func touchUp(_ ray: TouchRay, inside: Bool) {
        held = false
        grab.target = 0
    }

    func activate() { grab.kick(motion.kick(.grab)) }

    /// The stop the value sits on, with `stops`.
    var stopIndex: Int { Int((value * Double((stops ?? 101) - 1)).rounded()) }

    func adjust(by steps: Int) {
        set(((value + Double(steps) * step) / step).rounded() * step)
        grab.kick(motion.kick(.grab))
    }

    func update(_ dt: Double, _ ctx: UIContext) {
        presence.advance(dt, ctx)
        presence.apply(to: entity, rest: rest, reduceMotion: ctx.reduceMotion)
        guard entity.isEnabled else { return }
        if ctx.reduceMotion {
            knob.snap(to: value)
            grab.snap(to: held ? 1 : 0)
        } else {
            knob.advance(dt)
            grab.advance(dt)
        }
        let k = Float(min(1.08, max(-0.08, knob.value)))
        let x = -length / 2 + k * length
        puck.position.x = x
        let g = Float(grab.value)
        // Lean into the travel, squash while held.
        let lean = Float(max(-0.5, min(0.5, knob.velocity * 0.12)))
        puck.orientation = simd_quatf(angle: -lean, axis: [0, 0, 1])
        puck.scale = SIMD3(1 + 0.25 * g, 1 + 0.25 * g, 1 - 0.3 * g)
        let w = max(0.001, x + length / 2)
        fill.scale.x = w
        fill.position.x = -length / 2 + w / 2
    }
}
